src/Control/Effect/State/Lazy.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE DeriveFunctor, ExplicitForAll, FlexibleContexts, FlexibleInstances, KindSignatures, MultiParamTypeClasses, TypeOperators, UndecidableInstances #-}
+module Control.Effect.State.Lazy
+( State (..)
+, get
+, gets
+, put
+, modify
+, modifyLazy
+, runLazyState
+, evalLazyState
+, execLazyState
+, LazyStateC(..)
+) where
+
+import Control.Applicative (Alternative(..))
+import Control.Effect.Carrier
+import Control.Effect.Sum
+import Control.Effect.State
+import Control.Monad (MonadPlus(..))
+import Control.Monad.Fail
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import Prelude hiding (fail)
+
+-- | Replace the state value with the result of applying a function to the current state value.
+--   This is lazy in the new state; injudicious use of this function may lead to space leaks.
+modifyLazy :: (Member (State s) sig, Carrier sig m) => (s -> s) -> m ()
+modifyLazy f = get >>= put . f
```

Comment:
This is abstracted over the carrier, so I think it still makes sense in `Control.Effect.State` proper.